### PR TITLE
Attach icalendar invite for participants

### LINKED ISSUE #143: Attach iCalendar file (*.ics) to result e-mail
The result e-mail is currently very simple and it would be great if we could send an `*.ics` file along with it, so that people can just add the new meeting to their calendars. Integration with Indico will make this unnecessary in organizations using it, but it's still a nice thing to have in places where Indico is not an option.

image

### PR DESCRIPTION
**Pull request by saibot94**
Closes #143 .

I tested it out with gmail and it seems to parse the information properly.

image

This is the generated ICS file for the above, in case you want to take a look:

```
BEGIN:VCALENDAR
BEGIN:VEVENT
SUMMARY:Something like this
DTSTART;TZID=Europe/Paris;VALUE=DATE-TIME:20200914T110000
DTEND;TZID=Europe/Paris;VALUE=DATE-TIME:20200914T113000
DTSTAMP;VALUE=DATE-TIME:20200914T130547Z
ORGANIZER;CN="Cristi Sch":MAILTO:REDACTED
END:VEVENT
END:VCALENDAR
```

### CODE REVIEW

File: newdle/api.py [hunk 1]
```diff
@@ -399,6 +400,9 @@ def send_result_emails(code):
     date = newdle.final_dt.strftime('%-d %B %Y')
     start_time = newdle.final_dt.strftime('%H:%M')
     end_time = (newdle.final_dt + newdle.duration).strftime('%H:%M')
+    event = create_calendar_event(newdle)
+    # Attachments have to be in the format: (filename, content_bytes, MIMEType)
```
Comment (ThiefMaster) [inline newdle/api.py:404]: not sure if that comment is super useful here, it' kind of obvious that this is what the library expects..

if we want to document this, a better place would be the docstring of `notify_newdle_participants` which doesn't exist so for now i'd just not document it ;)

File: newdle/api.py [hunk 2]
```diff
@@ -399,6 +400,9 @@ def send_result_emails(code):
     date = newdle.final_dt.strftime('%-d %B %Y')
     start_time = newdle.final_dt.strftime('%H:%M')
     end_time = (newdle.final_dt + newdle.duration).strftime('%H:%M')
+    event = create_calendar_event(newdle)
```
Comment (ThiefMaster) [inline newdle/api.py:403]: i'd rename this to `ics` or `ical_data` - it's a string, not an ical Event object

File: newdle/calendar.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
+from datetime import datetime
+
+from flask import g
+from icalendar import Calendar, Event, vCalAddress, vText
+from pytz import timezone
+
+
+def create_calendar_event(newdle):
+    """Create an icalendar event based on a newdle"""
```
Comment (ThiefMaster) [inline newdle/calendar.py:9]:
```suggestion
    """Create an icalendar event based on a newdle."""
```

File: newdle/calendar.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    calendar = Calendar()
+    event = Event()
+    tz = timezone(newdle.timezone)
+    start_tz_aware = tz.localize(newdle.final_dt)
+    end_tz_aware = tz.localize(newdle.final_dt + newdle.duration)
```
Comment (ThiefMaster) [inline newdle/calendar.py:14]: This should be enough:

```suggestion
    end_tz_aware = start_tz_aware + newdle.duration
```

Comment (ThiefMaster) [reply newdle/calendar.py:14]: I'd just use `start_dt` and `end_dt` for these variable names

**Comment by ThiefMaster** [comment]
Would it make sense to include the list of participants in the ical event as well if the newdle is not set to private?

File: newdle/api.py (hunk 1 above)
Comment (saibot94) [reply newdle/api.py]: Thanks, removed it.

File: newdle/api.py (hunk 2 above)
Comment (saibot94) [reply newdle/api.py]: Changed to `ical_data`.

File: newdle/calendar.py (hunk 4 above)
Comment (saibot94) [reply newdle/calendar.py]: Changed

**Comment by saibot94** [comment]
> Would it make sense to include the list of participants in the ical event as well if the newdle is not set to private?

I think it's a good idea. Trying it out now.

File: newdle/calendar.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+from flask import g
+from icalendar import Calendar, Event, vCalAddress, vText
+from pytz import timezone
+
+
+def _email_to_vcal_address(email):
+    return vCalAddress(f'MAILTO:{email}')
+
+
+def add_participant_to_ical_event(event, participant):
```
Comment (ThiefMaster) [inline newdle/calendar.py:12]:
```suggestion
def _add_participant_to_ical_event(event, participant):
```

just like `_email_to_vcal_address` it's only used in here

File: newdle/calendar.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    event.add('dtstamp', datetime.utcnow())
+
+    organizer = _email_to_vcal_address(g.user['email'])
+    organizer.params['cn'] = vText(newdle.creator_name)
+    event['organizer'] = organizer
+
+    if not newdle.private:
+        participants = [p for p in newdle.participants if p.email]
+        for participant in participants:
+            add_participant_to_ical_event(event, participant)
```
Comment (ThiefMaster) [inline newdle/calendar.py:40]:
```suggestion
        for participant in newdle.participants:
            if not p.email:
                continue
            add_participant_to_ical_event(event, participant)
```

Your option is fine as well though so if you prefer that feel free to ignore this suggestion

Comment (saibot94) [reply newdle/calendar.py]: I think I prefer the first variant, thanks.

**Review by ThiefMaster — APPROVED** [review]
Very nice PR, thanks!

**Pull request merged**